### PR TITLE
[node]: Change Buffer variable declaration to declared interface instead of inline type

### PR DESCRIPTION
**Pull request by survirtual**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://nodejs.org/api/buffer.html#buffer_buffer
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

With the way Buffer was declared, it wasn't possible to add factory extension methods.  This PR corrects that.

For example, I want to create a factory extension method on Buffer, like so:

`Buffer.fromBase64 = function(b64: string): Buffer {
    return Buffer.from(b64, "base64");
};`

In order to add the type declaration for that, I'd want to do the following in my code normally:

`interface BufferConstructor {
    fromBase64: (b64: string) => Buffer;
}`

But since Buffer is declared as a global variable Buffer with an inline type, my extension method declaration wouldn't be valid prior to this PR.

### CODE REVIEW

**Comment by ZaneHannanAU** [comment]
… build issue, unrelated to the build itself.

https://api.travis-ci.org/v3/job/442849937/log.txt

```
Worker information
hostname: 1dd0e686-5db5-4348-8467-9dbaac1546e7@1.production-2-worker-org-08-packet
version: v4.0.0 https://github.com/travis-ci/worker/tree/e5cb567e10c0eefe380e81c9a2229ac8fb6a16ce
instance: 74f6b9d travisci/ci-garnet:packer-1512502276-986baf0 (via amqp)
startup: 7.069807217s


No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received

The build has been terminated
```

Also, iirc the `Buffer` is a per-instance one, so technically it wouldn't carry over.

Ideally you shouldn't be editing the constructors or w/e regardless so there's that.

**Comment by survirtual** [comment]
@ZaneHannanAU the per-instance `Buffer` is described prior, with `interface Buffer extends Uint8Array` at line 272.  The declaration modified in this PR is irt the static declaration.

In any case, making this change allowed me to write a static extension method without breaking `Buffer`, so I figured I'd share.

**Comment by ZaneHannanAU** [comment]
Cool. Now we just need to get the build to run through.

**Comment by survirtual** [comment]
Looks like some type definitions (or tests) need to be updated:

`Error: The following packages had errors: boom/v4, express-jsonschema, flux, karma-coverage, meteor, mongoose, multer-gridfs-storage, next, next-redux-saga, next-server, next-redux-wrapper, react, react-albus, react-flag-icon-css, react-helmet/v4, react-hot-loader, react-intl, react-is, react-jss, react-loadable, react-measure, react-native, react-native-indicators, react-native-multi-slider, react-native-modalbox, react-native-svg-charts, react-native-vector-icons, react-navigation, react-primitives, react-redux/v5, react-relay, react-select/v1, react-select, react-side-effect, react-stripe-elements, reactstrap, recompose, rrc, storybook-readme, styled-system, theming`

Most of the tests pass.  On first glance, the failures don't appear related to this PR.

File: types/node/index.d.ts [hunk 1]
```diff
@@ -476,7 +477,8 @@ declare const Buffer: {
      * This is the number of bytes used to determine the size of pre-allocated, internal Buffer instances used for pooling. This value may be modified.
      */
     poolSize: number;
-};
+}
+declare var Buffer: BufferConstructor;
```
Comment (ThomasdenH) [inline types/node/index.d.ts]: This should still be `const`.

**Review by ThomasdenH — CHANGES_REQUESTED** [review]
Sorry if it's a bit pedantic, but you changed a `const` declaration into a `var` declaration.

**Comment by SimonSchick** [comment]
As a general though, extending native types/globals is generally considered bad practise or even harmful.
I don't have any objections towards the change of the `Buffer` definition but the reason you are doing it is questionable.

File: types/node/index.d.ts (hunk 1 above)
Comment (survirtual) [reply types/node/index.d.ts]: It should!  Thank you.

**Comment by survirtual** [comment]
@SimonSchick I can understand that thought.  Whether or not extension methods are a good practice, this change just makes the Buffer type definition a bit more consistent with other types.

Though, I definitely do have an appreciation for extension methods in C#, and likewise here.  Being able to replicate that behavior and in some ways go beyond it within Typescript has been really useful for me in making my code simpler and more concise.

Is there a particular functional reason you consider this bad practice or is it more of a code style issue?

**Review by mohsen1 — APPROVED** [review]

**Comment by survirtual** [comment]
@ZaneHannanAU that does make sense, thank you for explaining.

I suppose factory functions do ultimately make me a bit uncomfortable as a standalone export, though I know it's the standard in the JS world and plays better with packages.  It just seems so detached from the domain of operation it exists in which can make it a bit trickier to reason about the code.  I get a lot of peace of mind being able to classify the factory methods according to the type they produce.

In any case, I appreciate you taking the time.

**Review by ThomasdenH — APPROVED** [review]

**Comment by survirtual** [comment]
@mohsen1 @ThomasdenH the build is failing because of projects dependent on these types unrelated to the changeset.  Can this PR get a policy override and force a merge?

**Pull request closed without merging**